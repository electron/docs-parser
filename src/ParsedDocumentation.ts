export enum DocumentationTag {
  OS_MACOS = 'os_macos',
  OS_MAS = 'os_mas',
  OS_WINDOWS = 'os_windows',
  OS_LINUX = 'os_linux',
  STABILITY_EXPERIMENTAL = 'stability_experimental',
  STABILITY_DEPRECATED = 'stability_deprecated',
  AVAILABILITY_READONLY = 'availability_readonly',
}
export declare type PossibleStringValue = {
  value: string;
  description: string;
};
export declare type DetailedStringType = {
  type: 'String';
  possibleValues: PossibleStringValue[] | null;
};
export declare type DetailedObjectType = {
  type: 'Object';
  properties: PropertyDocumentationBlock[];
};
export declare type DetailedFunctionType = {
  type: 'Function';
  parameters: MethodParameterDocumentation[];
  returns: TypeInformation | null;
};
export declare type DetailedType = (
  | {
      type: TypeInformation[];
    }
  | DetailedFunctionType
  | DetailedObjectType
  | DetailedStringType
  | {
      type: string;
    }
) & {
  innerTypes?: TypeInformation[];
};
export declare type TypeInformation = {
  collection: boolean;
} & DetailedType;
export declare type MethodParameterDocumentation = {
  name: string;
  description: string;
  required: boolean;
} & TypeInformation;
export declare type EventParameterDocumentation = {
  name: string;
  description: string;
  required: boolean;
} & TypeInformation;
export declare type DocumentationBlock = {
  name: string;
  description: string;
  additionalTags: DocumentationTag[];
};
export declare type MethodDocumentationBlock = DocumentationBlock & {
  signature: string;
  parameters: MethodParameterDocumentation[];
  returns: TypeInformation | null;
};
export declare type EventDocumentationBlock = DocumentationBlock & {
  parameters: EventParameterDocumentation[];
};
export declare type PropertyDocumentationBlock = DocumentationBlock & {
  required: boolean;
} & TypeInformation;
export declare type BaseDocumentationContainer = {
  name: string;
  extends?: string;
  description: string;
  version: string;
  slug: string;
  websiteUrl: string;
  repoUrl: string;
};
export declare type ProcessBlock = {
  main: boolean;
  renderer: boolean;
};
export declare type ModuleDocumentationContainer = {
  type: 'Module';
  process: ProcessBlock;
  methods: MethodDocumentationBlock[];
  events: EventDocumentationBlock[];
  properties: PropertyDocumentationBlock[];
  constructorMethod?: undefined;
  instanceMethods?: undefined;
  instanceEvents?: undefined;
  instanceProperties?: undefined;
  staticProperties?: undefined;
  staticMethods?: undefined;
} & BaseDocumentationContainer;
export declare type StructureDocumentationContainer = {
  type: 'Structure';
  properties: PropertyDocumentationBlock[];
  constructorMethod?: undefined;
  methods?: undefined;
  events?: undefined;
  instanceMethods?: undefined;
  instanceEvents?: undefined;
  instanceProperties?: undefined;
  staticProperties?: undefined;
  staticMethods?: undefined;
  extends?: string;
} & BaseDocumentationContainer;
export declare type ClassDocumentationContainer = {
  type: 'Class';
  process: ProcessBlock;
  constructorMethod: Pick<MethodDocumentationBlock, 'signature' | 'parameters'> | null;
  instanceName: string;
  staticMethods: MethodDocumentationBlock[];
  staticProperties: PropertyDocumentationBlock[];
  instanceMethods: MethodDocumentationBlock[];
  instanceEvents: EventDocumentationBlock[];
  instanceProperties: PropertyDocumentationBlock[];
  methods?: undefined;
  events?: undefined;
  properties?: undefined;
} & BaseDocumentationContainer;
export declare type ElementDocumentationContainer = {
  type: 'Element';
  process: ProcessBlock;
  constructorMethod?: undefined;
  methods: MethodDocumentationBlock[];
  events: EventDocumentationBlock[];
  properties: PropertyDocumentationBlock[];
  instanceMethods?: undefined;
  instanceEvents?: undefined;
  instanceProperties?: undefined;
  staticProperties?: undefined;
  staticMethods?: undefined;
} & BaseDocumentationContainer;
export declare type ParsedDocumentationResult = (
  | ModuleDocumentationContainer
  | ClassDocumentationContainer
  | StructureDocumentationContainer
  | ElementDocumentationContainer
)[];

export class ParsedDocumentation {
  private repr: ParsedDocumentationResult = [];

  public addStructure(struct: StructureDocumentationContainer) {
    this.repr.push(struct);
  }

  public addModuleOrClassOrElement(
    ...apiContainers: (
      | ModuleDocumentationContainer
      | ClassDocumentationContainer
      | ElementDocumentationContainer
    )[]
  ) {
    this.repr.push(...apiContainers);
  }

  public getJSON(): ParsedDocumentationResult {
    return this.repr.filter(container => {
      if (container.type !== 'Module') return true;

      return container.events.length + container.methods.length + container.properties.length > 0;
    });
  }
}
