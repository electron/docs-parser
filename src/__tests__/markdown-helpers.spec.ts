import * as fs from 'fs';
import * as path from 'path';
import MarkdownIt from 'markdown-it';

import {
  safelyJoinTokens,
  extractStringEnum,
  rawTypeToTypeInformation,
  parseHeadingTags,
} from '../markdown-helpers';
import { DocumentationTag } from '../ParsedDocumentation';

describe('markdown-helpers', () => {
  describe('parseHeadingTags', () => {
    it('should return an empty array for null input', () => {
      expect(parseHeadingTags(null)).toEqual([]);
    });

    it('should return an empty array if there are no tags in the input', () => {
      expect(parseHeadingTags('String thing no tags')).toEqual([]);
    });

    it('should return a list of tags if there is one tag', () => {
      expect(parseHeadingTags(' _macOS_')).toEqual([DocumentationTag.OS_MACOS]);
    });

    it('should return a list of tags if there are multiple tags', () => {
      expect(parseHeadingTags(' _macOS_ _Windows_ _Experimental_')).toEqual([
        DocumentationTag.OS_MACOS,
        DocumentationTag.OS_WINDOWS,
        DocumentationTag.STABILITY_EXPERIMENTAL,
      ]);
    });

    it('should throw an error if there is a tag not on the whitelist', () => {
      expect(() => parseHeadingTags(' _Awesome_')).toThrowErrorMatchingInlineSnapshot(
        `"heading tags must be from the whitelist: [\\"macOS\\",\\"mas\\",\\"Windows\\",\\"Linux\\",\\"Experimental\\",\\"Deprecated\\",\\"Readonly\\"]: expected [ Array(7) ] to include 'Awesome'"`,
      );
    });
  });

  describe('safelyJoinTokens', () => {
    it('should join no tokens to an empty string', () => {
      expect(safelyJoinTokens([])).toBe('');
    });

    describe('snapshots', () => {
      const fixtureDir = path.resolve(__dirname, 'fixtures');
      for (const markdownFixture of fs.readdirSync(fixtureDir)) {
        if (!markdownFixture.endsWith('.md')) continue;

        it(`should be correct for ${path.basename(markdownFixture, '.md')}`, () => {
          const markdown = new MarkdownIt();
          const tokens = markdown.parse(
            fs.readFileSync(path.resolve(fixtureDir, markdownFixture), 'utf8'),
            {},
          );
          expect(safelyJoinTokens(tokens)).toMatchSnapshot();
        });
      }
    });
  });

  describe('extractStringEnum()', () => {
    it('should return null if no string enum is found', () => {
      expect(
        extractStringEnum('this is just nonsense that can be anything really or include magic'),
      ).toBe(null);
    });

    it('should return null if nothing remotely string enum~ish exists', () => {
      expect(extractStringEnum('wassup')).toBe(null);
    });

    describe('with single quotes', () => {
      it('should extract an enum of the format "can be x"', () => {
        const values = extractStringEnum('Can be `x`')!;
        expect(values).not.toBe(null);
        expect(values).toHaveLength(1);
        expect(values[0].value).toBe('x');
      });

      it('should extract an enum of the format "can be x or y"', () => {
        const values = extractStringEnum('Can be `x` or `y`')!;
        expect(values).not.toBe(null);
        expect(values).toHaveLength(2);
        expect(values[0].value).toBe('x');
        expect(values[1].value).toBe('y');
      });

      it('should extract an enum of the format "can be x, y or z"', () => {
        const values = extractStringEnum('Can be `x`, `y` or `z`')!;
        expect(values).not.toBe(null);
        expect(values).toHaveLength(3);
        expect(values[0].value).toBe('x');
        expect(values[1].value).toBe('y');
        expect(values[2].value).toBe('z');
      });

      it('should extract an enum of the format "can be x, y, or z"', () => {
        const values = extractStringEnum('Can be `x`, `y`, or `z`')!;
        expect(values).not.toBe(null);
        expect(values).toHaveLength(3);
        expect(values[0].value).toBe('x');
        expect(values[1].value).toBe('y');
        expect(values[2].value).toBe('z');
      });

      it('should extract an enum of the format "values include a', () => {
        const values = extractStringEnum('Values include `a`')!;
        expect(values).not.toBe(null);
        expect(values).toHaveLength(1);
        expect(values[0].value).toBe('a');
      });

      it('should extract an enum of the format "values include a and b', () => {
        const values = extractStringEnum('Values include `a` and `b`')!;
        expect(values).not.toBe(null);
        expect(values).toHaveLength(2);
        expect(values[0].value).toBe('a');
        expect(values[1].value).toBe('b');
      });

      it('should extract an enum of the format "values include a, b and c', () => {
        const values = extractStringEnum('Values include `a`, `b` and `c`')!;
        expect(values).not.toBe(null);
        expect(values).toHaveLength(3);
        expect(values[0].value).toBe('a');
        expect(values[1].value).toBe('b');
        expect(values[2].value).toBe('c');
      });
    });

    describe('with backticks', () => {
      it('should extract an enum of the format "can be x"', () => {
        const values = extractStringEnum(`Can be 'x'`)!;
        expect(values).not.toBe(null);
        expect(values).toHaveLength(1);
        expect(values[0].value).toBe('x');
      });

      it('should extract an enum of the format "can be x or y"', () => {
        const values = extractStringEnum(`Can be 'x' or 'y'`)!;
        expect(values).not.toBe(null);
        expect(values).toHaveLength(2);
        expect(values[0].value).toBe('x');
        expect(values[1].value).toBe('y');
      });

      it('should extract an enum of the format "can be x, y or z"', () => {
        const values = extractStringEnum(`Can be 'x', 'y' or 'z'`)!;
        expect(values).not.toBe(null);
        expect(values).toHaveLength(3);
        expect(values[0].value).toBe('x');
        expect(values[1].value).toBe('y');
        expect(values[2].value).toBe('z');
      });

      it('should extract an enum of the format "can be x, y, or z"', () => {
        const values = extractStringEnum(`Can be 'x', 'y', or 'z'`)!;
        expect(values).not.toBe(null);
        expect(values).toHaveLength(3);
        expect(values[0].value).toBe('x');
        expect(values[1].value).toBe('y');
        expect(values[2].value).toBe('z');
      });

      it('should extract an enum of the format "values include a', () => {
        const values = extractStringEnum(`Values include 'a'`)!;
        expect(values).not.toBe(null);
        expect(values).toHaveLength(1);
        expect(values[0].value).toBe('a');
      });

      it('should extract an enum of the format "values include a and b', () => {
        const values = extractStringEnum(`Values include 'a' and 'b'`)!;
        expect(values).not.toBe(null);
        expect(values).toHaveLength(2);
        expect(values[0].value).toBe('a');
        expect(values[1].value).toBe('b');
      });

      it('should extract an enum of the format "values include a, b and c', () => {
        const values = extractStringEnum(`Values include 'a', 'b' and 'c'`)!;
        expect(values).not.toBe(null);
        expect(values).toHaveLength(3);
        expect(values[0].value).toBe('a');
        expect(values[1].value).toBe('b');
        expect(values[2].value).toBe('c');
      });
    });
  });

  describe('rawTypeToTypeInformation()', () => {
    it('should map a primitive types correctly', () => {
      expect(rawTypeToTypeInformation('Boolean', '', null)).toMatchSnapshot();
    });

    it('should map an unknown types correctly', () => {
      expect(rawTypeToTypeInformation('MyType', '', null)).toMatchSnapshot();
    });

    it('should map a Promise types correctly', () => {
      expect(rawTypeToTypeInformation('Promise<T>', '', null)).toMatchSnapshot();
    });

    it('should map a complex Promise types correctly', () => {
      expect(rawTypeToTypeInformation('Promise<T | A>', '', null)).toMatchSnapshot();
    });

    it('should map a nested Promise types correctly', () => {
      expect(rawTypeToTypeInformation('Promise<T | Promise<A>>', '', null)).toMatchSnapshot();
    });

    it('should map a nested complex Promise types correctly', () => {
      expect(rawTypeToTypeInformation('Promise<T | Promise<A[]>[]>', '', null)).toMatchSnapshot();
    });

    it('should map a nested complex Promise types correctly', () => {
      expect(rawTypeToTypeInformation('Promise<T | Promise<A[]>[]>', '', null)).toMatchSnapshot();
    });

    it('should map a nested Function types correctly', () => {
      expect(rawTypeToTypeInformation('Promise<T | Function<A[]>[]>', '', null)).toMatchSnapshot();
    });

    it('should map a wrapped collection type correctly', () => {
      expect(
        rawTypeToTypeInformation('Promise<(T | Function<A[]>)[]>', '', null),
      ).toMatchSnapshot();
    });

    it('should map a function return type correctly', () => {
      expect(rawTypeToTypeInformation('Function<R>', '', null)).toMatchSnapshot();
    });

    it('should map a function return type + param types correctly', () => {
      expect(rawTypeToTypeInformation('Function<P1, P2, R>', '', null)).toMatchSnapshot();
    });

    it('should map a function with complex return type + complex param types correctly', () => {
      expect(
        rawTypeToTypeInformation(
          'Function<P1<InnerP1, AnotherInnerP1[]>, P2<(InnerP2 | AnotherInnerP2<SuperDeepP2, EvenDeeperP2>)[]>, R<Foo>>',
          '',
          null,
        ),
      ).toMatchSnapshot();
    });
  });
});
