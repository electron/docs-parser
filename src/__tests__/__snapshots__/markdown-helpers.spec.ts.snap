// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdown-helpers rawTypeToTypeInformation() should map a Promise types correctly 1`] = `
Object {
  "collection": false,
  "innerTypes": Array [
    Object {
      "collection": false,
      "type": "T",
    },
  ],
  "type": "Promise",
}
`;

exports[`markdown-helpers rawTypeToTypeInformation() should map a complex Promise types correctly 1`] = `
Object {
  "collection": false,
  "innerTypes": Array [
    Object {
      "collection": false,
      "type": Array [
        Object {
          "collection": false,
          "type": "T",
        },
        Object {
          "collection": false,
          "type": "A",
        },
      ],
    },
  ],
  "type": "Promise",
}
`;

exports[`markdown-helpers rawTypeToTypeInformation() should map a function return type + param types correctly 1`] = `
Object {
  "collection": false,
  "parameters": Array [
    Object {
      "collection": false,
      "type": "P1",
    },
    Object {
      "collection": false,
      "type": "P2",
    },
  ],
  "returns": Object {
    "collection": false,
    "type": "R",
  },
  "type": "Function",
}
`;

exports[`markdown-helpers rawTypeToTypeInformation() should map a function return type correctly 1`] = `
Object {
  "collection": false,
  "parameters": Array [],
  "returns": Object {
    "collection": false,
    "type": "R",
  },
  "type": "Function",
}
`;

exports[`markdown-helpers rawTypeToTypeInformation() should map a function with complex return type + complex param types correctly 1`] = `
Object {
  "collection": false,
  "parameters": Array [
    Object {
      "collection": false,
      "innerTypes": Array [
        Object {
          "collection": false,
          "type": "InnerP1",
        },
        Object {
          "collection": true,
          "type": "AnotherInnerP1",
        },
      ],
      "type": "P1",
    },
    Object {
      "collection": false,
      "innerTypes": Array [
        Object {
          "collection": true,
          "type": Array [
            Object {
              "collection": false,
              "type": "InnerP2",
            },
            Object {
              "collection": false,
              "innerTypes": Array [
                Object {
                  "collection": false,
                  "type": "SuperDeepP2",
                },
                Object {
                  "collection": false,
                  "type": "EvenDeeperP2",
                },
              ],
              "type": "AnotherInnerP2",
            },
          ],
        },
      ],
      "type": "P2",
    },
  ],
  "returns": Object {
    "collection": false,
    "innerTypes": Array [
      Object {
        "collection": false,
        "type": "Foo",
      },
    ],
    "type": "R",
  },
  "type": "Function",
}
`;

exports[`markdown-helpers rawTypeToTypeInformation() should map a nested Function types correctly 1`] = `
Object {
  "collection": false,
  "innerTypes": Array [
    Object {
      "collection": false,
      "type": Array [
        Object {
          "collection": false,
          "type": "T",
        },
        Object {
          "collection": true,
          "parameters": Array [],
          "returns": Object {
            "collection": true,
            "type": "A",
          },
          "type": "Function",
        },
      ],
    },
  ],
  "type": "Promise",
}
`;

exports[`markdown-helpers rawTypeToTypeInformation() should map a nested Promise types correctly 1`] = `
Object {
  "collection": false,
  "innerTypes": Array [
    Object {
      "collection": false,
      "type": Array [
        Object {
          "collection": false,
          "type": "T",
        },
        Object {
          "collection": false,
          "innerTypes": Array [
            Object {
              "collection": false,
              "type": "A",
            },
          ],
          "type": "Promise",
        },
      ],
    },
  ],
  "type": "Promise",
}
`;

exports[`markdown-helpers rawTypeToTypeInformation() should map a nested complex Promise types correctly 1`] = `
Object {
  "collection": false,
  "innerTypes": Array [
    Object {
      "collection": false,
      "type": Array [
        Object {
          "collection": false,
          "type": "T",
        },
        Object {
          "collection": true,
          "innerTypes": Array [
            Object {
              "collection": true,
              "type": "A",
            },
          ],
          "type": "Promise",
        },
      ],
    },
  ],
  "type": "Promise",
}
`;

exports[`markdown-helpers rawTypeToTypeInformation() should map a nested complex Promise types correctly 2`] = `
Object {
  "collection": false,
  "innerTypes": Array [
    Object {
      "collection": false,
      "type": Array [
        Object {
          "collection": false,
          "type": "T",
        },
        Object {
          "collection": true,
          "innerTypes": Array [
            Object {
              "collection": true,
              "type": "A",
            },
          ],
          "type": "Promise",
        },
      ],
    },
  ],
  "type": "Promise",
}
`;

exports[`markdown-helpers rawTypeToTypeInformation() should map a primitive types correctly 1`] = `
Object {
  "collection": false,
  "type": "Boolean",
}
`;

exports[`markdown-helpers rawTypeToTypeInformation() should map a wrapped collection type correctly 1`] = `
Object {
  "collection": false,
  "innerTypes": Array [
    Object {
      "collection": true,
      "type": Array [
        Object {
          "collection": false,
          "type": "T",
        },
        Object {
          "collection": false,
          "parameters": Array [],
          "returns": Object {
            "collection": true,
            "type": "A",
          },
          "type": "Function",
        },
      ],
    },
  ],
  "type": "Promise",
}
`;

exports[`markdown-helpers rawTypeToTypeInformation() should map an unknown types correctly 1`] = `
Object {
  "collection": false,
  "type": "MyType",
}
`;

exports[`markdown-helpers safelyJoinTokens snapshots should be correct for basic-paragraph 1`] = `"This is just a basic paragraph.  It has multiple sentences and natural soft breaks."`;

exports[`markdown-helpers safelyJoinTokens snapshots should be correct for blockquotes 1`] = `
"This is a paragraph

> and here is a quote

and another paragraph"
`;

exports[`markdown-helpers safelyJoinTokens snapshots should be correct for list-after-paragraph 1`] = `
"This is a paragraph

* this
* is
* a
* list"
`;

exports[`markdown-helpers safelyJoinTokens snapshots should be correct for lists 1`] = `
"* bar
* thing
* stuff
* my tab key does not work and I am sad"
`;

exports[`markdown-helpers safelyJoinTokens snapshots should be correct for multiple-paragraphs 1`] = `
"This is paragraph 1, it has a few sentences.  Like this one, and a soft break.

And now this is a second paragraph, and it's also quite short with a soft break."
`;

exports[`markdown-helpers safelyJoinTokens snapshots should be correct for nested-list 1`] = `
"* top level
  * second level
* back to top
  * back to second
  * still at second
    * now at third"
`;

exports[`markdown-helpers safelyJoinTokens snapshots should be correct for paragraph-fence-removal 1`] = `
"This fence should be removed

Hey"
`;

exports[`markdown-helpers safelyJoinTokens snapshots should be correct for paragraph-with-inline-code 1`] = `"This is a \`inline code\` block that is \`code\` tagged."`;

exports[`markdown-helpers safelyJoinTokens snapshots should be correct for paragraph-with-links 1`] = `"This paragraph has a bunch of stuff.  Also some links Foo, these links should be stripped."`;

exports[`markdown-helpers safelyJoinTokens snapshots should be correct for paragraph-with-text-markers 1`] = `"Heya, this is **bold** and _italic_ and *italic* and ~~struck through~~"`;
